## Phase 2: Variant Discovery

This should be run on the entire batch of reads (samples), provided that the batch is not too large (more on that below)

CWL: [exomeseq-02-variantdiscovery.cwl](exomeseq-02-variantdiscovery.cwl)

Questions:

- How many samples can we expect?
- If user does not have enough samples, how do we feed additional samples in for joint genotyping
- What confidence and sensitivety to use
- Recalibration annotations must differ based on the type of analysis being done (e.g. population with >10 samples or family with 3 samples)

Datasets (from b37):

- SNP resources: Hapmap, omni, 1000genomes
- DBSNP
- indels from Mills

Tasks:

- Update example JSON - Have many development JSON files from different steps

Inputs:

- intervals: `File[]?` - genomic interval over which to operate. optional.
- mapped_reads: `File[]` - array of input samples, in BAM format from pre-processing step, with indexes
- reference\_genome: `File` (with secondaryFiles of the bwa index.) Using bwa 0.7.x, 64 bit indexes are default and we do NOT use the `.64` file suffix.
  - `b37/human_g1k_v37.fasta`
- threads: `int?` - Number of threads to use. This is disabled in a lot of the GATK tool descriptions because the tools fail with it.
- GATKJar: `File` - the GATK Jar file (must be supplied at runtime)
- stand\_call\_conf: `double` - the standard calling confidence threshold for calling a variant
- snp\_resource\_hapmap: `File` - VCF file SNP locations from hapmap
  - `b37/hapmap/hapmap_3.3.b37.vcf`
- snp\_resource\_omni: `File` - VCF file of SNP locations from omni
  - `b37/omni/1000G_omni2.5.b37.vcf`
- snp\_resource\_1kg: `File` - VCF file of SNP locations from 1000 genomes
  - `b37/1000G_phase1.snps.high_confidence.b37.vcf`
- resource\_dbsnp: `File` - VCF file of SNP locations from dbSNP
  - `b37/dbsnp_138.b37.vcf`
- indel\_resource\_mills: `File` - VCF file of indel locations from Mills and 1000G
  - `b37/Mills_and_1000G_gold_standard.indels.b37.vcf`

Outputs:

- per\_sample\_raw\_variants: `File[]` - Output of HaplotypeCaller variant calls for each input sample
- joint\_raw\_variants: `File` - Single VCF file from Joint Genotyping (from above)
- variant\_recalibration\_snps\_tranches: `File` - The output tranches file used by ApplyRecalibration in SNP mode
- variant\_recalibration\_snps\_recal: `File` - The output recal file used by ApplyRecalibration in SNP mode
- variant\_recalibration\_snps\_rscript: `File`- The output rscript file generated by the VQSR in SNP mode to aid in visualization of the input data and learned model
- variant\_recalibration\_indels\_tranches: `File`- The output tranches file used by ApplyRecalibration in INDEL mod
- variant\_recalibration\_indels\_recal: `File`- The output recal file used by ApplyRecalibration in INDEL mode
- variant\_recalibration\_indels\_rscript: `File`- The output rscript file generated by the VQSR in INDEL mode to aid in visualization of the input data and learned model
- variant\_recalibration\_snps\_vcf: `File`- The output filtered and recalibrated VCF file in SNP mode in which each variant is annotated with its VQSLOD value
- variant\_recalibration\_indels\_vcf: `File`- The output filtered and recalibrated VCF file in INDEL mode in which each variant is annotated with its VQSLOD value

Steps:

- Runs GATK HaplotypeCaller on each input BAM file to call variants, producing gvcf files(Should merge after if working with more than a few hundred samples). See **Merge (Optional)** at [Variant Discovery](https://software.broadinstitute.org/gatk/best-practices/bp_3step.php?case=GermShortWGS&p=2)
- Performs joint genotyping on all gvcf files
- Recalibrates SNP variants
- Recalibrates Indel variants
-
