#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: Workflow
requirements:
  - class: ScatterFeatureRequirement
inputs:
  # For testing, intervals on reference genome
  intervals: string[]?
  # reference genome, fasta
  reference_genome: File
  # Number of threads to use
  threads: int?
  # GATK
  GATKJar: File
  # Variant Recalibration - SNPs
  snp_resource_hapmap: File
  snp_resource_omni: File
  snp_resource_1kg: File
  # Variant Recalibration - Common
  resource_dbsnp: File
  # Variant Recalibration - Indels
  indel_resource_mills: File

  # Just for testing
  variants: File

outputs:
  variant_recalibration_snps_tranches:
    type: File
    outputSource: variant_recalibration_snps/tranches_File
    doc: "The output tranches file used by ApplyRecalibration in SNP mode"
  variant_recalibration_snps_recal:
    type: File
    outputSource: variant_recalibration_snps/recal_File
    doc: "The output recal file used by ApplyRecalibration in SNP mode"
  variant_recalibration_snps_rscript:
    type: File
    outputSource: variant_recalibration_snps/vqsr_rscript
    doc: "The output rscript file generated by the VQSR in SNP mode to aid in visualization of the input data and learned model"
  variant_recalibration_indels_tranches:
    type: File
    outputSource: variant_recalibration_indels/tranches_File
    doc: "The output tranches file used by ApplyRecalibration in INDEL mode"
  variant_recalibration_indels_recal:
    type: File
    outputSource: variant_recalibration_indels/recal_File
    doc: "The output recal file used by ApplyRecalibration in INDEL mode"
  variant_recalibration_indels_rscript:
    type: File
    outputSource: variant_recalibration_indels/vqsr_rscript
    doc: "The output rscript file generated by the VQSR in INDEL mode to aid in visualization of the input data and learned model"
  variant_recalibration_snps_vcf:
    type: File
    outputSource: apply_recalibration_snps/output_recalibrated_vcf
    doc: "The output filtered and recalibrated VCF file in SNP mode in which each variant is annotated with its VQSLOD value"
  variant_recalibration_indels_vcf:
    type: File
    outputSource: apply_recalibration_indels/output_recalibrated_vcf
    doc: "The output filtered and recalibrated VCF file in INDEL mode in which each variant is annotated with its VQSLOD value"
steps:
  variant_recalibration_snps:
    run: ../community-workflows/tools/GATK-VariantRecalibrator-SNPs.cwl
    in:
      GATKJar: GATKJar
      intervals: intervals
      reference: reference_genome
      variants: variants
      threads:
        default: 1
      outputfile_recal:
        default: "snps_vqsr_recal.out"
      outputfile_tranches:
        default: "snps_vqsr_tranches.out"
      outputfile_rscript:
        default: "snps_vqsr.R"
      resource_hapmap: snp_resource_hapmap
      resource_omni: snp_resource_omni
      resource_1kg: snp_resource_1kg
      resource_dbsnp: resource_dbsnp
      # annotations, See https://software.broadinstitute.org/gatk/documentation/article?id=1259
      # Removed DP and InbreedingCoeff
      # Please note that these recommendations are formulated for whole-genome datasets.
      # For exomes, we do not recommend using DP for variant recalibration (see below for details of why).
      annotations:
        # The InbreedingCoeff is a population level statistic that requires at least 10 samples in order to be computed. For projects with fewer samples, or that includes many closely related samples (such as a family) please omit this annotation from the command line.
        # NOTE: InbreedingCoeff would need to be annotated separately
        default: ["QD","MQ","MQRankSum","ReadPosRankSum","FS","SOR"]
    out:
      - tranches_File
      - recal_File
      - vqsr_rscript
  apply_recalibration_snps:
    run: ../community-workflows/tools/GATK-ApplyRecalibration.cwl
    in:
      GATKJar: GATKJar
      intervals: intervals
      reference: reference_genome
      variants: variants
      threads: threads
      tranches_file: variant_recalibration_snps/tranches_File
      recal_file: variant_recalibration_snps/recal_File
      outputfile_recalibrated_vcf:
        default: "snps_recalibrated.vcf"
      ts_filter_level:
        default: 99.5
      mode:
        default: "SNP"
    out:
      - output_recalibrated_vcf
  variant_recalibration_indels:
    run: ../community-workflows/tools/GATK-VariantRecalibrator-Indels.cwl
    in:
      GATKJar: GATKJar
      intervals: intervals
      reference: reference_genome
      variants: variants
      threads:
        default: 1
      outputfile_recal:
        default: "indels_vqsr_recal.out"
      outputfile_tranches:
        default: "indels_vqsr_tranches.out"
      outputfile_rscript:
        default: "indels_vqsr.R"
      resource_mills: indel_resource_mills
      resource_dbsnp: resource_dbsnp
      # annotations, See https://software.broadinstitute.org/gatk/documentation/article?id=1259
      # Removed DP and InbreedingCoeff
      # Please note that these recommendations are formulated for whole-genome datasets.
      # For exomes, we do not recommend using DP for variant recalibration (see below for details of why).
      annotations:
        # The InbreedingCoeff is a population level statistic that requires at least 10 samples in order to be computed. For projects with fewer samples, or that includes many closely related samples (such as a family) please omit this annotation from the command line.
        # NOTE: InbreedingCoeff would need to be annotated separately
                default: ["QD","FS","SOR","ReadPosRankSum","MQRankSum"]
    out:
      - tranches_File
      - recal_File
      - vqsr_rscript
  apply_recalibration_indels:
    run: ../community-workflows/tools/GATK-ApplyRecalibration.cwl
    in:
      GATKJar: GATKJar
      intervals: intervals
      reference: reference_genome
      variants: variants
      threads: threads
      tranches_file: variant_recalibration_indels/tranches_File
      recal_file: variant_recalibration_indels/recal_File
      outputfile_recalibrated_vcf:
        default: "indels_recalibrated.vcf"
      ts_filter_level:
        default: 99.0
      mode:
        default: "INDEL"
    out:
      - output_recalibrated_vcf
